File: apps/moot-mgr/Sources/MootManager/ControlChannel.swift
// ControlChannel.swift
//
// The gated control channel for moot-mgr (P3) over a Unix domain socket.
//
// ============================ SECURITY BOUNDARY =============================
// Privileged state changes (monitoring on/off, set retention) travel this UDS,
// NOT the loopback HTTP surface. The socket file is created at mode 0600
// (owner-only read/write — enforced in POSIXSocket.listenUnix). A browser
// cannot speak a Unix domain socket — that is the feature (concepts §1.6):
// privileged ops stay off the loopback HTTP path by construction, and the
// filesystem permission bits (0600) are the access gate. No token is needed on
// the UDS because the OS already authenticates by the process's effective uid
// via the socket-file permission bits.
//
// The protocol is line-oriented and tiny: a client connects, writes ONE control
// path optionally followed by a tab and a JSON body, terminated by a newline;
// the server applies the verb and writes back the JSON ControlResult. Verbs are
// the SAME control paths the HTTP control surface uses, dispatched through the
// shared HTTPReadAPI.applyControl so both surfaces have identical semantics.
//
//   request : "/api/control/monitoring/on\n"
//   request : "/api/control/retention\t{\"seconds\":3600}\n"
//   response: "{\"detail\":\"...\",\"ok\":true}\n"
// ===========================================================================
//
// Implementation: a POSIX Unix-domain listening socket (POSIXSocket.swift). NO
// external packages. (NWListener cannot bind a server UDS at a chosen path in
// this environment; the POSIX path binds and chmods 0600 directly.) The accept
// loop runs on a dedicated thread; each connection is served on the actor.

import Foundation
import OSLog

#if canImport(Glibc)
import Glibc
#else
import Darwin
#endif

// MARK: - ControlChannel

/// The UDS-backed gated control channel. One instance per resident host.
public actor ControlChannel {

    /// The API whose `applyControl` implements the verbs (shared with the HTTP
    /// control surface so both behave identically).
    private let api: HTTPReadAPI

    /// Filesystem path of the Unix domain socket.
    private let socketPath: String

    private let logger = Logger(subsystem: "com.mootx01.kit", category: "ControlChannel")

    private var listenFD: Int32 = -1
    private var acceptThread: Thread?
    private let running = RunningFlag()

    /// The mode the socket file is created/verified at: owner-only (0600).
    static let socketMode: mode_t = 0o600

    // MARK: - Init

    /// Create the control channel.
    ///
    /// - Parameters:
    ///   - api:        The read-API whose `applyControl` implements the verbs.
    ///   - socketPath: Filesystem path for the UDS. Any existing file here is
    ///                 removed on `start()` so a stale socket cannot be reused.
    public init(api: HTTPReadAPI, socketPath: String) {
        self.api = api
        self.socketPath = socketPath
    }

    // MARK: - Lifecycle

    /// Create the socket (bound + chmod 0600) and begin accepting connections.
    ///
    /// - Throws: `SocketError` if the socket cannot be created/bound/chmod'd.
    public func start() throws {
        let fd = try POSIXSocket.listenUnix(path: socketPath)
        self.listenFD = fd
        running.set(true)

        let thread = Thread { [weak self, fd, running] in
            while running.get() {
                guard let cfd = POSIXSocket.acceptOne(fd) else {
                    if !running.get() { break }
                    continue
                }
                guard let self else { close(cfd); break }
                Task { await self.serve(cfd) }
            }
        }
        thread.name = "com.mootx01.kit.ControlChannel.accept"
        thread.start()
        self.acceptThread = thread
        logger.info("ControlChannel listening on UDS \(self.socketPath) (0600)")
    }

    /// Stop accepting and remove the socket file. Idempotent.
    public func stop() async {
        running.set(false)
        if listenFD >= 0 {
            close(listenFD)
            listenFD = -1
        }
        acceptThread = nil
        unlink(socketPath)
    }

    // MARK: - Connection handling

    /// Read one request line, dispatch the verb, write the JSON result.
    private func serve(_ fd: Int32) async {
        defer { close(fd) }
        guard let line = readLine(fd) else { return }
        // Split on the first tab: path before, optional JSON body after.
        let (path, body): (String, Data) = {
            if let tab = line.firstIndex(of: "\t") {
                let p = String(line[line.startIndex..<tab])
                let b = String(line[line.index(after: tab)...])
                return (p.trimmingCharacters(in: .whitespacesAndNewlines), Data(b.utf8))
            }
            return (line.trimmingCharacters(in: .whitespacesAndNewlines), Data())
        }()

        let response = await api.applyControl(path: path, body: body)
        var out = response.json
        out.append(Data("\n".utf8))
        POSIXSocket.sendAll(fd, out)
    }

    /// Read until the first newline (the request terminator), returning the line
    /// without the newline. Caps the read to guard against an unbounded request.
    private func readLine(_ fd: Int32) -> String? {
        var buffer = Data()
        let cap = 64 * 1024
        while true {
            if let nl = buffer.firstIndex(of: 0x0A) {
                return String(data: buffer[buffer.startIndex..<nl], encoding: .utf8)
            }
            if buffer.count > cap { return nil }
            guard let chunk = POSIXSocket.recv(fd, max: 16 * 1024), !chunk.isEmpty else {
                // Connection closed; return whatever line we have, if any.
                return buffer.isEmpty ? nil : String(data: buffer, encoding: .utf8)
            }
            buffer.append(chunk)
        }
    }
}

File: apps/moot-mgr/rust/src/control_channel.rs
// control_channel.rs — Rust twin of the Swift moot-mgr ControlChannel.swift.
//
// ============================ SECURITY BOUNDARY =============================
// Privileged state changes (monitoring on/off, set retention, estate provision/
// lifecycle) travel a local IPC channel, NOT the loopback HTTP surface. A
// browser cannot speak that channel — that is the feature: privileged ops stay
// off the loopback HTTP path by construction, and the OS authenticates the
// connecting peer for us:
//
//   * Unix (Linux/macOS): a filesystem Unix-domain socket created at mode 0600
//     (owner-only read/write). The permission bits ARE the access gate — the OS
//     authenticates by the connecting process's effective uid. No token needed.
//   * Windows: a named pipe carrying the creating user's default ACL (owner-only).
//     The ACL is the equivalent gate; the pipe is not on the filesystem, so its
//     name is derived from the control-socket path (see `windows_pipe_name`).
//
// The protocol is line-oriented and tiny: a client connects, writes ONE control
// path optionally followed by a tab and a JSON body, terminated by a newline;
// the server applies the verb and writes back the JSON ControlResponse. The
// verbs are the SAME control paths the HTTP control surface uses, dispatched
// through the shared `HttpReadApi::apply_control` so both surfaces have identical
// semantics.
//
//   request : "/api/control/monitoring/on\n"
//   request : "/api/control/retention\t{\"seconds\":3600}\n"
//   response: "{\"detail\":\"...\",\"ok\":true}\n"
// ===========================================================================
//
// Implementation: a cross-platform local socket via the `interprocess` crate —
// a Unix-domain socket on Unix (chmod'd 0600 directly) and a named pipe on
// Windows. The accept loop runs on a dedicated thread; each connection is served
// inline. Per the platform law the Rust vertical targets Windows AND Linux, so
// the transport must compile and run on both.

use std::io::{Read, Write};
#[cfg(unix)]
use std::os::unix::fs::PermissionsExt;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};
use std::thread::JoinHandle;

use interprocess::local_socket::prelude::*;
use interprocess::local_socket::{Listener, ListenerNonblockingMode, ListenerOptions, Stream};

use crate::http_read_api::HttpReadApi;

/// The mode the Unix socket file is created/verified at: owner-only (0600).
/// Mirrors Swift `ControlChannel.socketMode`. (Windows uses the named-pipe ACL
/// instead — see the security-boundary header.)
#[cfg(unix)]
const SOCKET_MODE: u32 = 0o600;

/// Read cap to guard against an unbounded request line (matches the Swift 64 KiB).
const MAX_REQUEST_BYTES: usize = 64 * 1024;

/// The gated control channel. One instance per resident host. Mirrors Swift
/// `ControlChannel`.
pub struct ControlChannel {
    /// The API whose `apply_control` implements the verbs (shared with the HTTP
    /// control surface so both behave identically).
    api: Arc<HttpReadApi>,
    /// On Unix, the filesystem path of the Unix-domain socket (any existing file
    /// here is removed on `start()` so a stale socket cannot be reused). On
    /// Windows, the seed from which the named-pipe identifier is derived.
    socket_path: String,
    running: Arc<AtomicBool>,
    accept_thread: Mutex<Option<JoinHandle<()>>>,
}

impl ControlChannel {
    /// Create the control channel. Mirrors Swift `ControlChannel.init(api:socketPath:)`.
    pub fn new(api: Arc<HttpReadApi>, socket_path: impl Into<String>) -> Self {
        ControlChannel {
            api,
            socket_path: socket_path.into(),
            running: Arc::new(AtomicBool::new(false)),
            accept_thread: Mutex::new(None),
        }
    }

    /// Create the listener (UDS chmod 0600 on Unix / owner-ACL named pipe on
    /// Windows) and begin accepting connections on a dedicated thread. Mirrors
    /// Swift `ControlChannel.start()`.
    pub fn start(&self) -> std::io::Result<()> {
        let listener = bind_listener(&self.socket_path)?;
        // Nonblocking accept: the loop polls `running` between accepts instead of
        // blocking indefinitely. This makes shutdown deterministic on BOTH
        // platforms without a wakeup connection — critical on Windows, where a
        // named-pipe "poke" races the accept and can leave it blocked forever
        // (see the platform note on `Listener::accept`).
        listener.set_nonblocking(ListenerNonblockingMode::Accept)?;
        self.running.store(true, Ordering::SeqCst);

        let api = Arc::clone(&self.api);
        let running = Arc::clone(&self.running);
        let socket_path = self.socket_path.clone();
        let handle = std::thread::Builder::new()
            .name("moot-mgr.ControlChannel.accept".to_string())
            .spawn(move || {
                for stream in listener.incoming() {
                    if !running.load(Ordering::SeqCst) {
                        break;
                    }
                    match stream {
                        Ok(s) => serve(&api, s),
                        // No client waiting (nonblocking). Sleep briefly, then
                        // loop — re-checking `running` so stop() takes effect
                        // within one tick.
                        Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
                            std::thread::sleep(std::time::Duration::from_millis(20));
                        }
                        Err(_) => {
                            if !running.load(Ordering::SeqCst) {
                                break;
                            }
                        }
                    }
                }
                // Best-effort cleanup of the Unix socket file on loop exit so a
                // restart binds cleanly (no-op on Windows, where the named pipe
                // is released when the listener drops).
                let _ = std::fs::remove_file(&socket_path);
            })?;
        *self.accept_thread.lock().unwrap() = Some(handle);
        Ok(())
    }

    /// Stop accepting and remove the socket file. Idempotent. Mirrors Swift
    /// `ControlChannel.stop()`.
    ///
    /// The accept loop polls `running` between nonblocking accepts, so flipping
    /// the flag is enough — it exits within one poll tick on its own. No wakeup
    /// connection is needed (and none would be reliable on Windows named pipes).
    pub fn stop(&self) {
        self.running.store(false, Ordering::SeqCst);
        if let Some(handle) = self.accept_thread.lock().unwrap().take() {
            let _ = handle.join();
        }
        let _ = std::fs::remove_file(&self.socket_path);
    }
}

/// Bind the gated control-channel listener for this platform.
///
/// Unix: a filesystem Unix-domain socket at `socket_path`, chmod'd 0600 — the
/// owner-only permission bits ARE the access gate. Windows: a named pipe whose
/// name is derived from `socket_path`, carrying the creating user's default ACL
/// (owner-only). A browser can speak neither, by construction.
fn bind_listener(socket_path: &str) -> std::io::Result<Listener> {
    #[cfg(unix)]
    {
        use interprocess::local_socket::GenericFilePath;
        // Remove a stale socket file so the bind does not fail on EADDRINUSE.
        let _ = std::fs::remove_file(socket_path);
        let name = socket_path.to_fs_name::<GenericFilePath>()?;
        let listener = ListenerOptions::new().name(name).create_sync()?;
        // chmod 0600 — owner-only. THIS is the Unix access gate (the OS
        // authenticates by the connecting process's uid via these bits).
        std::fs::set_permissions(
            socket_path,
            std::fs::Permissions::from_mode(SOCKET_MODE),
        )?;
        Ok(listener)
    }
    #[cfg(windows)]
    {
        use interprocess::local_socket::GenericNamespaced;
        let pipe = windows_pipe_name(socket_path);
        let name = pipe.to_ns_name::<GenericNamespaced>()?;
        // Windows access gate: `interprocess` v2 creates the named pipe with the
        // calling process's default DACL (Windows default), which restricts access
        // to the pipe owner and members of the Administrators group. This provides
        // the owner-only equivalent of the Unix 0600 chmod — remote/arbitrary
        // users on the same machine cannot connect.
        //
        // Explicit DACL control (e.g. restricting to exactly the creating SID with
        // no Administrators inheritance) would require a direct `windows-sys` call
        // to `CreateNamedPipeW` with a custom SECURITY_ATTRIBUTES. That would
        // introduce a new external dependency, which the no-new-external-dep rule
        // prohibits. The `interprocess` v2 API does not expose DACL configuration
        // through `ListenerOptions`. The default DACL is the correct hardening
        // posture for the local manager: it excludes unprivileged local users,
        // which is the threat model. Administrator-level access is expected to have
        // equivalent privilege to the moot-mgr process itself.
        ListenerOptions::new().name(name).create_sync()
    }
}

/// Derive a stable Windows named-pipe identifier from the control-socket path.
/// Named pipes are not filesystem objects, so the per-data-dir control-socket
/// PATH is folded (FNV-1a) into a flat-namespace name: distinct data dirs get
/// distinct pipes; the same data dir always maps to the same pipe.
#[cfg(windows)]
fn windows_pipe_name(socket_path: &str) -> String {
    let mut h: u64 = 0xcbf2_9ce4_8422_2325;
    for b in socket_path.as_bytes() {
        h ^= u64::from(*b);
        h = h.wrapping_mul(0x0000_0100_0000_01b3);
    }
    format!("mootx01-mgr-{h:016x}.sock")
}

/// Read one request line, dispatch the verb through the shared `apply_control`,
/// write the JSON result + newline. Mirrors Swift `ControlChannel.serve(_:)`.
fn serve(api: &HttpReadApi, mut stream: Stream) {
    let line = match read_line(&mut stream) {
        Some(l) => l,
        None => return,
    };
    // Split on the first tab: path before, optional JSON body after.
    let (path, body): (String, Vec<u8>) = match line.find('\t') {
        Some(tab) => {
            let p = line[..tab].trim().to_string();
            let b = line[tab + 1..].as_bytes().to_vec();
            (p, b)
        }
        None => (line.trim().to_string(), Vec::new()),
    };

    let response = api.apply_control(&path, &body);
    let mut out = response.json;
    out.push(b'\n');
    let _ = stream.write_all(&out);
    let _ = stream.flush();
}

/// Read until the first newline (the request terminator), returning the line
/// without the newline. Caps the read against an unbounded request. Mirrors
/// Swift `ControlChannel.readLine(_:)`.
fn read_line(stream: &mut Stream) -> Option<String> {
    let mut buffer: Vec<u8> = Vec::new();
    let mut chunk = [0u8; 16 * 1024];
    loop {
        if let Some(nl) = buffer.iter().position(|&b| b == b'\n') {
            return Some(String::from_utf8_lossy(&buffer[..nl]).into_owned());
        }
        if buffer.len() > MAX_REQUEST_BYTES {
            return None;
        }
        let n = stream.read(&mut chunk).ok()?;
        if n == 0 {
            // Connection closed; return whatever line we have, if any.
            return if buffer.is_empty() {
                None
            } else {
                Some(String::from_utf8_lossy(&buffer).into_owned())
            };
        }
        buffer.extend_from_slice(&chunk[..n]);
    }
}
